Reduce by key, extends #1141

biddisco opened the pull request:
This is an algorithm present in thrust:: but not included in N4409. It is useful for a number of algorithms that use stream compaction (common in visualization).

hkaiser commented on hpx/parallel/algorithms/reduce_by_key.hpp:
```diff
@@ -0,0 +1,501 @@
+    }
+
+    // for debugging
+    std::ostream &operator<<(std::ostream &os,
```
I'd like to see this being hidden behind a pp constant. Also, if this is enabled, it requires `<iostreams>` to be included.

hkaiser commented:
May I ask to change the naming convention partially used (for instance for `ReduceKeySeriesStates`, `ReduceStencilGeneration` and various variable names)? For this project we have settled to use `gnu_style_names` for types, variables, and functions. We use `CamelStyle` for template parameters only. `ALL_CAPS` is used for macro names. We don't use Hungarian Notation either (prefixing variable names with its shortened type (i.e. `fStart`, `fEnd`).

I'm sorry to bring this up as it might seem unrelated and demeaning to the work you are doing. We try however to keep the coding style in the library as uniform as possible.

hkaiser commented:
If the new algorithm is meant to be used for non-asynchronous execution policies only (no `task`) I'd suggest to prevent it from compiling when invoked with one. This should also be documented.

OTOH, creating a simple wrapper for the case of asynchronous policies which launches the corresponding non-asynchronous version in a new thread would do the trick as well. This might also simplify the implementation of the non-asynchronous version.

hkaiser commented:
You added extensive doxygen comments to enable the creation of the documentation for the new algorithms. In order for the documentation to be actually created though, you'd need to add the file to the build system. Several sports would need to be touched:

- you might need to add a doxygen `\file` directive to the implementation file
- add the file here for it to be parsed by doxygen: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/CMakeLists.txt#L63
- add the algorithm names here for it to show up in the index: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/hpx.idx#L144
- we have a couple of tables listing all implemented algorithms here: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/manual/parallel_algorithms.qbk#L119, you algorithm should be added to the appropriate table

Let me know if you would like to have some help with this.

biddisco commented:
> unrelated and demeaning
Indeed. However, I will grudgingly comply with the naming - and fix the doc generation.
The async execution I'll need to look at more closely.

biddisco commented:
I've fixed the async execution and corrected (I believe) all the offending variable names, docs etc.
This was rebased onto the latest master today, so once the tests have passed I would be happy to see it merged.

hkaiser commented on hpx/parallel/algorithms/reduce_by_key.hpp:
```diff
@@ -0,0 +1,561 @@
+            (boost::is_base_of<std::random_access_iterator_tag, category2>::value) ||
+            (boost::is_base_of<std::output_iterator_tag, category3>::value) ||
+            (boost::is_base_of<std::output_iterator_tag, category4>::value),
+            "iterators : Random_access for inputs and Output for outputs.");
```
Did you intend to use `&&` instead of `||`? Also, any iterator type except input_iterators can be used as an output iterator. What you probably want to write in order to check whether an iterator conforms to the output iterator concept is:

    hpx::traits::is_output_iterator<OutIter>::value || 
        hpx::traits::is_forward_iterator<OutIter>::value 

We also have a 

    hpx::traits::is_random_access_iterator<RanIter>::value

which could be used instead (this also gets rid of the boost/type_traits facilities which we replaced everywhere else in the parallelism files)..

hkaiser commented on tests/unit/parallel/algorithms/reduce_by_key.cpp:
```diff
@@ -0,0 +1,323 @@
+    template<typename T>
+    void output(const std::string &name, const std::vector<T> &v) {
+#ifdef EXTRA_DEBUG
+        std::cout << name.c_str() << "\t : {" << v.size() << "} : ";
```
If `EXTRA_DEBUG` is enabled you miss an `#include <iostreams>`.
